README.md
# docker-mock

[![Build Status](https://img.shields.io/travis/Runnable/docker-mock/master.svg?style=flat-square)](https://travis-ci.org/Runnable/docker-mock)
[![Dependency Status](https://img.shields.io/david/Runnable/docker-mock.svg?style=flat-square)](https://david-dm.org/Runnable/docker-mock)
[![devDependency Status](https://img.shields.io/david/dev/Runnable/docker-mock.svg?style=flat-square)](https://david-dm.org/Runnable/docker-mock#info=devDependencies)
[![Code Climate](https://img.shields.io/codeclimate/github/Runnable/docker-mock.svg?style=flat-square)](https://codeclimate.com/github/Runnable/docker-mock)
[![Test Coverage](https://img.shields.io/codeclimate/coverage/github/Runnable/docker-mock.svg?style=flat-square)](https://codeclimate.com/github/Runnable/docker-mock)

[![NPM](https://nodei.co/npm/docker-mock.png?compact=true)](https://nodei.co/npm/docker-mock/)

A mock for [Docker](http://docker.io)!

## Usage

Require it from your node program:

```javascript
var dockerMock = require('docker-mock');
dockerMock.listen(5354);
```

You can also use the command line interface, just run `docker-mock` after installing the package globally.

## Examples

See the tests for some sample usage, but you should be able to point your docker client at this mock and test against it.

## Failures

This does have support for simulating failures during build. This is done (using `dockerode`) by doing the following:

```javascript
// file is a tar containing at minimum a Dockerfile
var file = ...;
docker.buildImage(
  file,
  {
    t: 'doomedImage',
    fail: true
  },
  function (err, res) {
    // err will not be null
  });
```


## Configuration

If you want to disable randomly generated events exposed under `/events` endpoint please use ENV var: `DISABLE_RANDOM_EVENTS=true`.


## Events

You can manually emit docker mock events.

```javascript
  var dockerMock = require('docker-mock');
  dockerMock.listen(5354);
  dockerMock.events.stream.emit('data', JSON.stringify({ status: 'die', from: '..', id: '...', time: '...' }));
  // or
  dockerMock.events.stream.emit('data', dockerMock.events.generateEvent());

```

## Contributing

Please make sure all unit tests pass, lint passes, and coverage remains high during development (see below for details).

### Testing

Testing is done locally via `npm test`.

### Coverage

Coverage is now run by Lab via `npm test`. Output at the bottom shows percent coverage verses a threshold set in `package.json`.


CHANGELOG.md
## Change Log

### v5.3.0 (2016/01/02 22:13 +00:00)
- [#57](https://github.com/Runnable/docker-mock/pull/57) Fixing and Expanding Label Functionality (@bkendall)
- [#56](https://github.com/Runnable/docker-mock/pull/56) cleanup: apply formatter (@bkendall)

### v5.2.0 (2015/12/25 17:56 +00:00)
- [#55](https://github.com/Runnable/docker-mock/pull/55) Handle name when creating a container and returning list of containers (@doganyazar)
- [#54](https://github.com/Runnable/docker-mock/pull/54) Standard (@bkendall)
- [#53](https://github.com/Runnable/docker-mock/pull/53) SemiStandard (@bkendall)

### v5.1.1 (2015/11/01 01:57 +00:00)
- [#51](https://github.com/Runnable/docker-mock/pull/51) Update Deps; Mocha + Chai (@bkendall)
- [#50](https://github.com/Runnable/docker-mock/pull/50) travis badge for master (@bkendall)

### v5.1.0 (2015/10/21 23:35 +00:00)
- [#49](https://github.com/Runnable/docker-mock/pull/49) control exit code via kill signals (@tjmehta)

### v5.0.2 (2015/10/14 21:26 +00:00)
- [#48](https://github.com/Runnable/docker-mock/pull/48) fix: move callback-count to dependencies (@bkendall)

### v5.0.1 (2015/10/09 01:28 +00:00)
- [#47](https://github.com/Runnable/docker-mock/pull/47) fixed NetworkSettings start bug (@tjmehta)

### v5.0.0 (2015/10/08 23:33 +00:00)
- [#46](https://github.com/Runnable/docker-mock/pull/46) fix inspect containerName value (@tjmehta)

### v4.0.0 (2015/10/06 22:09 +00:00)
- [#42](https://github.com/Runnable/docker-mock/pull/42) Promises (@bkendall)

### v3.0.1 (2015/10/06 20:20 +00:00)
- [#45](https://github.com/Runnable/docker-mock/pull/45) fix get by name (@tjmehta)
- [#44](https://github.com/Runnable/docker-mock/pull/44) travis: iojs -> 4 (@bkendall)

### v3.0.0 (2015/10/06 18:17 +00:00)
- [#43](https://github.com/Runnable/docker-mock/pull/43) fix capitalization issue for container name (@tjmehta)

### v2.2.0 (2015/08/27 21:17 +00:00)
- [#39](https://github.com/Runnable/docker-mock/pull/39) add load and save to docker (@anandkumarpatel, @bkendall)
- [#41](https://github.com/Runnable/docker-mock/pull/41) lint updates (@bkendall)
- [#40](https://github.com/Runnable/docker-mock/pull/40) Save/Load Cleanup (@bkendall)
- [#36](https://github.com/Runnable/docker-mock/pull/36) HTTPS Test (@bkendall)
- [#38](https://github.com/Runnable/docker-mock/pull/38) Dependencies Update (@bkendall)

### v2.1.0 (2015/06/25 00:32 +00:00)
- [#37](https://github.com/Runnable/docker-mock/pull/37) GET /images/json respond w/ Created property + tests (@cflynn07)

### v2.0.0 (2015/05/12 16:45 +00:00)
- [#35](https://github.com/Runnable/docker-mock/pull/35) docker 1.6 Labels support (@cflynn07)
- [#34](https://github.com/Runnable/docker-mock/pull/34) coverage update (@bkendall)
- [#33](https://github.com/Runnable/docker-mock/pull/33) update deps, add coverage (@bkendall)

### v1.0.2 (2015/04/28 07:22 +00:00)
- [#32](https://github.com/Runnable/docker-mock/pull/32) ESLint, Cleanup (@bkendall)

### v1.0.1 (2015/03/12 18:50 +00:00)
- [#30](https://github.com/Runnable/docker-mock/pull/30) /events fixes (@tjmehta, @bkendall)
- [#31](https://github.com/Runnable/docker-mock/pull/31) fixes for event fixes (@bkendall)
- [#29](https://github.com/Runnable/docker-mock/pull/29) add node 0.12, use containers in travis (@bkendall)

### v1.0.0 (2015/02/24 01:58 +00:00)
- [#20](https://github.com/Runnable/docker-mock/pull/20) Code climate badge (@podviaznikov)
- [#26](https://github.com/Runnable/docker-mock/pull/26) Lab update, middleware logic fix (@bkendall)

### v0.6.1 (2015/02/16 23:58 +00:00)
- [#25](https://github.com/Runnable/docker-mock/pull/25) container destroy event (@tjmehta)

### v0.6.0 (2015/01/29 17:59 +00:00)
- [#24](https://github.com/Runnable/docker-mock/pull/24) Image History Endpoint (@bkendall)

### v0.5.7 (2015/01/15 06:15 +00:00)
- [#23](https://github.com/Runnable/docker-mock/pull/23) support container routes by name (@tjmehta)

### v0.5.5 (2015/01/07 23:21 +00:00)
- [#21](https://github.com/Runnable/docker-mock/pull/21) adding 443 as a port in to expose (@bkendall)

### v0.5.3 (2014/12/09 22:09 +00:00)
- [#17](https://github.com/Runnable/docker-mock/pull/17) add image pulling (@anandkumarpatel)
- [#18](https://github.com/Runnable/docker-mock/pull/18) Stop, restart, kill should emit several events (@podviaznikov)

### v0.5.2 (2014/11/30 06:20 +00:00)
- [#16](https://github.com/Runnable/docker-mock/pull/16) remove file requirement from docker-logs-mock (@tjmehta)
- [#13](https://github.com/Runnable/docker-mock/pull/13) Feature emit events from mocks (@podviaznikov)

### v0.5.1 (2014/11/19 23:38 +00:00)
- [#15](https://github.com/Runnable/docker-mock/pull/15) Create container middleware change (@cflynn07)

### v0.4.0 (2014/11/17 22:03 +00:00)
- [#11](https://github.com/Runnable/docker-mock/pull/11) add events endpoint (@podviaznikov)

### v0.3.1 (2014/11/13 19:34 +00:00)
- [#10](https://github.com/Runnable/docker-mock/pull/10) better restart and kill mocks (@tjmehta)

### v0.2.1 (2014/11/04 22:45 +00:00)
- [#7](https://github.com/Runnable/docker-mock/pull/7) more accurate start and stop (@tjmehta)

### v0.1.4 (2014/09/25 18:39 +00:00)
- [#6](https://github.com/Runnable/docker-mock/pull/6) adding support for env value on create container (@bkendall)

### v0.1.0 (2014/07/18 00:16 +00:00)
- [#5](https://github.com/Runnable/docker-mock/pull/5) add Pid to dock inspect (@anandkumarpatel)

lib/models/container.js
'use strict'

var assign = require('101/assign')
var EventEmitter = require('events').EventEmitter
var mapKeys = require('object-loops/map-keys')
var NotModifiedError = require('./base-store').NotModifiedError
var Promise = require('bluebird')
var util = require('util')
var utils = require('../utils')

module.exports = Container

function Container (opts) {
  if (!opts) { opts = {} }
  var Id = utils.randomId()
  var data = {
    Id: Id,
    Memory: 0,
    MemorySwap: 0,
    Image: null,
    Config: {
      AttachStderr: true,
      AttachStdin: false,
      AttachStdout: true,
      Cmd: [],
      Env: [],
      ExposedPorts: {},
      Hostname: '',
      Image: null,
      // FIXME: this is a hack
      Labels: opts.Labels || {},
      OpenStdin: false,
      PortSpecs: null,
      StdinOnce: false,
      Tty: false,
      User: '',
      Volumes: {},
      WorkingDir: ''
    },
    HostConfig: {
      Dns: null,
      VolumesFrom: ''
    },
    Volumes: {},
    State: {
      Running: false,
      Pid: -1
    },
    NetworkSettings: {
      Bridge: '',
      Gateway: '',
      IPAddress: '',
      IPPrefixLen: 0,
      MacAddress: '',
      Ports: null
    }
  }
  delete opts.Labels
  var capitalizedOpts = mapKeys(opts, utils.capitalize)
  if (capitalizedOpts.Name) {
    capitalizedOpts.Name = '/' + capitalizedOpts.Name
  }
  assign(this, data, capitalizedOpts)
}
util.inherits(Container, EventEmitter)

Container.prototype.start = function (wasRestart) {
  return Promise.resolve()
    .bind(this)
    .then(function () {
      if (this.State.Running) {
        throw new NotModifiedError('Container already running')
      }

      this.State = {
        // TODO: more data!
        Running: true,
        Pid: utils.newPid()
      }
      this.NetworkSettings = {
        Bridge: 'docker0',
        Gateway: '172.17.42.1',
        IPAddress: '172.17.0.' + utils.newPort(),
        IPPrefixLen: 16,
        MacAddress: '02:42:ac:11:00:05',
        Ports: {
          '80/tcp': [{ HostPort: utils.newPort() }],
          '443/tcp': [{ HostPort: utils.newPort() }],
          '15000/tcp': [{ HostPort: utils.newPort() }]
        }
      }
      return this
    })
    .then(function (container) {
      this.emit('event', 'start', container)
      if (wasRestart) {
        this.emit('event', 'restart', container)
      }
      return container
    })
}

Container.prototype.stop = function (ignoreStopped, killSignal) {
  var exitCode = (killSignal === 'SIGKILL') ? 1 : 0
  return Promise.resolve()
    .bind(this)
    .then(function () {
      if (this.State.Running === false && !ignoreStopped) {
        throw new NotModifiedError('Container already stopped')
      }

      this.State = {
        // TODO: more data!
        ExitCode: exitCode,
        Running: false,
        Pid: 0
      }
      this.NetworkSettings = {
        Bridge: '',
        Gateway: '',
        IPAddress: '',
        IPPrefixLen: 0,
        MacAddress: '',
        Ports: null
      }

      return this
    })
    .then(function (container) {
      this.emit('event', 'die', container)
      if (!ignoreStopped) {
        // stop doesn't ignore stopped
        this.emit('event', 'stop', container)
      } else if (ignoreStopped === 'kill') {
        this.emit('event', 'kill', container)
      }
      // nothing special for restart
      return container
    })
}
